### PR TITLE
tests, network: Refactor services test

### PR DESCRIPTION
**Pull request by EdDev**
### What this PR does

Refactor the e2e network services tests and helpers.

This is part of a bigger goal to cover all network tests by the linter.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/network/services.go [hunk 1]
```diff
@@ -239,85 +139,113 @@ var _ = SIGDescribe("Services", func() {
...
+			const subdomain = "vmi"
+			const hostname = "inbound"
+
+			inboundVMI = libvmi.NewFedora(append(
+				libvmi.WithMasqueradeNetworking(),
+				libvmi.WithLabel(selectorLabelKey, selectorLabelValue),
+				libvmi.WithSubdomain(subdomain),
+				libvmi.WithHostname(hostname),
+			)...)
+			var err error
```
Comment (dankenigsberg) [inline tests/network/services.go:152]: please educate me: why is this better than having `inboundVMI, err :=` below?

File: tests/network/services.go [hunk 2]
```diff
@@ -239,85 +139,113 @@ var _ = SIGDescribe("Services", func() {
...
+
+				By("setting up resources to expose the VMI via a service")
+				if ipFamily == k8sv1.IPv6Protocol {
+					serviceName += "v6"
+					service = netservice.BuildIPv6Spec(serviceName, servicePort, servicePort, selectorLabelKey, selectorLabelValue)
+				} else {
+					service = netservice.BuildSpec(serviceName, servicePort, servicePort, selectorLabelKey, selectorLabelValue)
+				}
+
+				c := kubevirt.Client()
```
Comment (dankenigsberg) [inline tests/network/services.go:175]: I prefer the shortform `c`, but everywhere else in this file you have `virtClient`. This inconsistency may perplex the reader.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
Thank you very much for this PR. I could not wrap my head around those lambdas.

I have few minor questions.

File: tests/network/services.go [hunk 3]
```diff
@@ -38,7 +38,6 @@ import (
 	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	v1 "kubevirt.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (EdDev) [reply tests/network/services.go:40]: Done

File: tests/network/services.go (hunk 1 above)
Comment (dankenigsberg) [reply tests/network/services.go:147]: :-( I cannot say that I like GoLang.

File: tests/network/services.go [hunk 4]
```diff
@@ -225,18 +192,32 @@ var _ = SIGDescribe("Services", func() {
...
 		})
 	})
 })
 
+func deferServiceCleanup(namespace, name string) {
+	DeferCleanup(func() {
+		err := kubevirt.Client().CoreV1().Services(namespace).Delete(context.Background(), name, k8smetav1.DeleteOptions{})
+		Expect(err).To(SatisfyAny(
+			Not(HaveOccurred()),
+			MatchError(errors.IsNotFound, "does not exist"),
```
[the remark this answers is not in the document]
Comment (dankenigsberg) [reply tests/network/services.go:202]: > we do not really care if they are already gone

I would contest this. If the job is miraculously gone, it means that the test is doing something unexpected. Tests should be written in a deterministic way, and fail if they deviate from our expectation.

Anyway, that's not related to your PR.

File: tests/network/services.go [hunk 5]
```diff
@@ -50,182 +49,81 @@ import (
...
+				job, err := runConnectivityJobToService(serviceName, inboundVMI.Namespace, servicePort)
 				Expect(err).NotTo(HaveOccurred(), expectConnectivityToExposedService)
+				deferJobCleanup(job.Namespace, job.Name)
 			})
 
 			It("[test_id:1548] should fail to reach the vmi if an invalid servicename is used", func() {
-				var err error
-
-				jobCleanup, err = assertNoConnectivityToService("wrongservice", inboundVMI.Namespace, servicePort)
+				job, err := runNoConnectivityJobToService("wrongservice", inboundVMI.Namespace, servicePort)
```
Comment (dankenigsberg) [inline tests/network/services.go:102]: Going back to deferred cleanup: I think that this test code would be more tight if you:
1. create job
2. defer its cleanup
3. expect that it fails

You would not allow a 1-3-2 order in production code; you shouldn't have it in tests either.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/network/services.go (hunk 5 above)
Comment (EdDev) [reply tests/network/services.go:102]: Now that this was explained offline, I get it.

The problem here is that `runNoConnectivity..` is doing two things: Created the job and waits for it to finish. We should defer the cleanup if it was created, regardless if the "wait" succeeded or failed.
In the scenario that the "wait" failed, the explicit cleanup will not be registered, which is wrong.

Fixed.

File: tests/network/services.go [hunk 6]
```diff
@@ -291,33 +256,47 @@ var _ = SIGDescribe("Services", func() {
...
 
-				jobCleanup, err = assertNoConnectivityToService(serviceName, inboundVMI.Namespace, servicePort)
+				err = job.WaitForJobToFail(tcpJob, 89*time.Second)
 				Expect(err).NotTo(HaveOccurred(), "connectivity is *not* expected, since there isn't an exposed service")
 			})
 		})
 	})
 })
+
+func createConnectivityJobToService(serviceName string, namespace string, servicePort int) (*batchv1.Job, error) {
```
Comment (dankenigsberg) [inline tests/network/services.go:229]: Not sure about the English, but how about `createServiceConnectivityJob` ?

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/network/services.go (hunk 6 above)
Comment (EdDev) [reply tests/network/services.go]: Done

**Review by dankenigsberg — APPROVED** [review]

File: tests/network/services.go [hunk 7]
```diff
@@ -58,21 +58,6 @@ const (
 var _ = SIGDescribe("Services", func() {
 	var virtClient kubecli.KubevirtClient
 
-	cleanupVMI := func(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) {
```
Comment (AlonaKaplan) [inline tests/network/services.go:86]: Please mention in the commit message that the VMI cleanup was dropped.
As already mentioned - I would drop the jo and service clean as well. The general test cleanup will take care of it anyway.

**Comment by EdDev** [comment]
change: Rebase

**Review by EdDev — COMMENTED** [review]
Thank you for the review.

I have added a commit at the end to remove the service & job deletions that appear in the test files and depend on the general deletion one.

I am not sure about the stability of doing this, I left it as a separated commit so we can revert it if needed. If you prefer to squash it in the commit that extract the deletion to helpers, let me know.

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/4a7195fd177013231668b43bdd5bed62c670148f..8412724eced2d0550ad8510551cfba3348359898): Answered review from @AlonaKaplan .

File: tests/network/services.go [hunk 8]
```diff
@@ -166,44 +129,49 @@ var _ = SIGDescribe("Services", func() {
...
+				DeferCleanup(func() {
+					err := virtClient.BatchV1().Jobs(tcpJob.Namespace).Delete(context.Background(), tcpJob.Name, k8smetav1.DeleteOptions{})
+					Expect(err).To(SatisfyAny(
+						Not(HaveOccurred()),
+						MatchError(errors.IsNotFound, "does not exist"),
+					))
+				})
 
-				jobCleanup, err = assertNoConnectivityToService("wrongservice", inboundVMI.Namespace, servicePort)
+				err = job.WaitForJobToFail(tcpJob, 89*time.Second)
```
[the remark this answers is not in the document]
Comment (dankenigsberg) [reply tests/network/services.go]: (I thought it was to silence a silly linter)

**Comment by EdDev** [comment]
change: Rebase

**Review by dankenigsberg — APPROVED** [review]

**Comment by AlonaKaplan** [comment]
/approve
Thanks!

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/48a5ab2455a71bb98c69f38994a9cb0b819ee4fd..3f556b6ede03c5287d1c363646827ea897f12ed9): Fixed a bug that probably got introduced through the rebases.

**Review by dankenigsberg — APPROVED** [review]

**Comment by EdDev** [comment]
[Flaky](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11187/pull-kubevirt-unit-test/1760206311804375040) unit test, unrelated to this change.

/test pull-kubevirt-unit-test

**Comment by EdDev** [comment]
We have SLIRP failures of VMI timeout.
I guess it is due to load. It has nothing to do with the changes in this PR.

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11187/pull-kubevirt-e2e-k8s-1.27-sig-network/1760259648096571392

/test pull-kubevirt-e2e-k8s-1.27-sig-network

**Pull request merged**